src/systems/filecoin_blockchain/storage_power_consensus/_index.md

```diff
@@ -44,54 +46,126 @@ More precisely,
+// updated in future. 
+GetBeaconEntryForEpoch(epoch) []BeaconEntry {
+    var rand Randomness
+    for rand == nil {
+        seeds := GetBeaconEntriesForEpoch(epoch)
+        if seeds == nil {
+            epoch -= 1
+        } else {
+            rand = seeds[len(seeds)-1].Randomness
+        }
+    }
+    return rand
+}
+
+GetBeaconEntriesForEpoch(epoch) []BeaconEntry {
+    maxDrandRound := MaxBeaconRoundForEpoch(epoch)
+
+    if epoch == 0 {
+        // special case genesis
+        entries := []
+        rand := drand.Public(curr)
+        return append(entries, rand)
+    }
+
+    prevMaxDrandRound := MaxBeaconRoundForEpoch(epoch - 1)
+    if (maxDrandRound == prevMaxDrandRound) {
+        // no new beacon randomness
+        return nil
+    }
+
+    entries := []
+    curr := maxDrandRound
+    // fetch all drand entries generated between this epoch and last
+    for curr > prevMaxDrandRound {
+        rand := drand.Public(curr)
+        entries = append(entries, rand)
+        curr -= 1
+    }
+    return entries
+}
+```
 
-- The {{<sref sector_sealer>}} uses tickets as SealSeeds to bind sector commitments to a given subchain.
-- The {{<sref post_generator>}} likewise uses tickets as PoStChallenges to prove sectors remain committed as of a given block.
-- They are drawn by the Storage Power subsystem as randomness in {{<sref leader_election>}} to determine their eligibility to mine a block
-- They are drawn by the Storage Power subsystem in order to generate new tickets for future use.
+{{<label beacon_entry_validation>}}
+### Validating Beacon Entries on block reception
+
+Per the above, a Filecoin chain will contain the entirety of the beacon's output from the Filecoin genesis to the current block.
+
+Given their role in leader election and other critical protocols in Filecoin, a block's beacon entries must be validated for every block. See {{<sref drand>}} for details. This can be done by ensuring every beacon entry is a valid signature over the prior one in the chain, using drand's [`Verify`](https://github.com/drand/drand/blob/master/beacon/beacon.go#L72) endpoint as follows:
+
+```go
+// This need not be done for the genesis block
+// We assume that blockHeader and priorBlockHeader are two valid subsequent headers where block was mined atop priorBlock
+ValidateBeaconEntries(blockHeader, priorBlockHeader) error {
+    currEntries := blockHeader.BeaconEntries
+    prevEntries := priorBlockHeader.BeaconEntries
+
+    var entries []BeaconEntry
+    // at currIdx == 0, must fetch last Fil block's last BeaconEntry
+    entries := append(entries, prevEntries[len(prevEntries)-1])
+    entries := append(entries, currEntries...)
+
+    currIdx := len(entries) - 1
+    for currIdx >= 0 {
+        // walking back the entries to ensure they follow one another
+        currEntry := entries[currIdx]
+        prevEntry := entries[currIdx - 1]
+        err := drand.Verify(node.drandPubKey, prevEntry.Data, currEntry.Data, currEntry.Round)
```

Comment:
It can't be in the future.

Reply:
Attack:
- The current block should have entries from X till Y, instead the miner adds entries from X until Y+1
- This function verifies all the entries X till Y+1 and this check will pass.

Reply:
Fix: check that currEntry.Round <= MaxBeaconRoundForEpoch(epoch)

Reply:
@nicola check out https://github.com/filecoin-project/lotus/blob/master/chain/beacon/beacon.go#L30. They do a slightly diff check. First they check that if the last block includes the `maxBeaconEntry` for this block's epoch, then this block has no beacon entries (i.e. that entries cannot be replayed on the chain), second they check exactly that `currEntry.Round == MaxBeaconRoundForEpoch`, not `<=`. So their checks are more restrictive than ours. Lmk which approach you prefer, adapted to theirs for now.

Reply:
Following their way will work!